src/test/jtx/impl/AMM.cpp

```diff
@@ -0,0 +1,787 @@
+{
+    if (isXRP(a))
+        return a.xrp();
+    return a;
+}
+
+static IOUAmount
+initialTokens(STAmount const& asset1, STAmount const& asset2)
+{
+    auto const product = number(asset1) * number(asset2);
+    return (IOUAmount)(
+        product.mantissa() >= 0 ? root2(product) : root2(-product));
+}
+
+AMM::AMM(
+    Env& env,
+    Account const& account,
+    STAmount const& asset1,
+    STAmount const& asset2,
+    bool log,
+    std::uint32_t tfee,
+    std::uint32_t fee,
+    std::optional<std::uint32_t> flags,
+    std::optional<jtx::seq> seq,
+    std::optional<jtx::msig> ms,
+    std::optional<ter> const& ter)
+    : env_(env)
+    , creatorAccount_(account)
+    , asset1_(asset1)
+    , asset2_(asset2)
+    , initialLPTokens_(initialTokens(asset1, asset2))
+    , ter_(ter)
+    , log_(log)
+    , lastPurchasePrice_(0)
+    , minSlotPrice_(0)
+    , bidMin_()
+    , bidMax_()
+    , msig_(ms)
+    , fee_(fee)
+{
+    create(tfee, flags, seq);
+}
+
+AMM::AMM(
+    Env& env,
+    Account const& account,
+    STAmount const& asset1,
+    STAmount const& asset2,
+    ter const& ter,
+    bool log)
+    : AMM(env,
+          account,
+          asset1,
+          asset2,
+          log,
+          0,
+          0,
+          std::nullopt,
+          std::nullopt,
+          std::nullopt,
+          ter)
+{
+}
+
+void
+AMM::create(
+    std::uint32_t tfee,
+    std::optional<std::uint32_t> flags,
+    std::optional<jtx::seq> seq)
+{
+    Json::Value jv;
+    jv[jss::Account] = creatorAccount_.human();
+    jv[jss::Amount] = asset1_.getJson(JsonOptions::none);
+    jv[jss::Amount2] = asset2_.getJson(JsonOptions::none);
+    jv[jss::TradingFee] = tfee;
+    jv[jss::TransactionType] = jss::AMMCreate;
+    if (flags)
+        jv[jss::Flags] = *flags;
+    if (fee_ != 0)
+        jv[jss::Fee] = std::to_string(fee_);
+    else
+        jv[jss::Fee] = std::to_string(env_.current()->fees().increment.drops());
+    submit(jv, seq, ter_);
+
+    if (!ter_ || env_.ter() == tesSUCCESS)
+    {
+        if (auto const amm = env_.current()->read(
+                keylet::amm(asset1_.issue(), asset2_.issue())))
+        {
+            ammAccount_ = amm->getAccountID(sfAccount);
+            lptIssue_ = ripple::ammLPTIssue(
+                asset1_.issue().currency,
+                asset2_.issue().currency,
+                ammAccount_);
+        }
+    }
+}
+
+std::optional<Json::Value>
+AMM::ammRpcInfo(
+    std::optional<AccountID> const& account,
+    std::optional<std::string> const& ledgerIndex,
+    std::optional<std::pair<Issue, Issue>> tokens) const
+{
+    Json::Value jv;
+    if (account)
+        jv[jss::account] = to_string(*account);
+    if (ledgerIndex)
+        jv[jss::ledger_index] = *ledgerIndex;
+    if (tokens)
+    {
+        jv[jss::asset] =
+            STIssue(sfAsset, tokens->first).getJson(JsonOptions::none);
+        jv[jss::asset2] =
+            STIssue(sfAsset2, tokens->first).getJson(JsonOptions::none);
```

Comment:
`tokens->second`. It looks like the only caller that provides `tokens` uses the same value for both, which is why tests didn't fail.